lark/parsers/lalr_puppet.py

```diff
@@ -1,107 +1,55 @@
 # This module provide a LALR puppet, which is used to debugging and error handling
 
-from copy import deepcopy
+from copy import copy
 
 from .lalr_analysis import Shift, Reduce
 from .. import Token
-from ..exceptions import ParseError
+from ..exceptions import UnexpectedToken
 
 
 class ParserPuppet(object):
     """ParserPuppet gives you advanced control over error handling when parsing with LALR.
 
     For a simpler, more streamlined interface, see the ``on_error`` argument to ``Lark.parse()``.
     """
-    def __init__(self, parser, state_stack, value_stack, start, stream, set_state):
+    def __init__(self, parser, parser_state, lexer_state):
         self.parser = parser
-        self._state_stack = state_stack
-        self._value_stack = value_stack
-        self._start = start
-        self._stream = stream
-        self._set_state = set_state
-
-        self.result = None
+        self.parser_state = parser_state
+        self.lexer_state = lexer_state
 
     def feed_token(self, token):
         """Feed the parser with a token, and advance it to the next state, as if it recieved it from the lexer.
 
         Note that ``token`` has to be an instance of ``Token``.
         """
-        end_state = self.parser.parse_table.end_states[self._start]
-        state_stack = self._state_stack
-        value_stack = self._value_stack
-
-        state = state_stack[-1]
-        action, arg = self.parser.parse_table.states[state][token.type]
-        if arg == end_state:
-            raise ParseError(arg)
-
-        while action is Reduce:
-            rule = arg
-            size = len(rule.expansion)
-            if size:
-                s = value_stack[-size:]
-                del state_stack[-size:]
-                del value_stack[-size:]
-            else:
-                s = []
-
-            value = self.parser.callbacks[rule](s)
-
-            _action, new_state = self.parser.parse_table.states[state_stack[-1]][rule.origin.name]
-            assert _action is Shift
-            state_stack.append(new_state)
-            value_stack.append(value)
-
-            if state_stack[-1] == end_state:
-                self.result = value_stack[-1]
-                return self.result
-
-            state = state_stack[-1]
-            try:
-                action, arg = self.parser.parse_table.states[state][token.type]
-            except KeyError as e:
-                raise ParseError(e)
-            assert arg != end_state
-
-        assert action is Shift
-        state_stack.append(arg)
-        value_stack.append(token)
-
-    def copy(self):
+        return self.parser_state.feed_token(token)
+
+    def __copy__(self):
```

Comment:
This is backwards compatibility breaking, since we exposed `ParserPuppet` with the `copy` method

Reply:
I agree, but it's a very new feature, so I doubt anyone had a chance to use it yet.

But okay, I'll keep the `copy()` method.